src/counts.jl

```diff
@@ -283,6 +283,31 @@ function addcounts_dict!(cm::Dict{T}, x::AbstractArray{T}) where T
     return cm
 end
 
+# If the bits type is of small size i.e. it can have up to 65536 distinct values
+# then it is always better to apply a counting-sort like reduce algorithm for 
+# faster results and less memory usage. However we still wish to enable others
+# to write generic algorithms, therefore the two methods below still accept the 
+# `alg` argument but it is ignored.
+function addcounts!(cm::Dict{Bool}, x::AbstractArray{Bool}; alg = :ignored)
+    sumx = sum(x)
+    cm[true] = get(cm, true, 0) + sumx
+    cm[false] = get(cm, false, 0) + length(x) - sumx
+    cm
+end
+
+function addcounts!(cm::Dict{T}, x::Vector{T}; alg = :ignored) where T <: Union{UInt8, UInt16, Int8, Int16}
+    counts = zeros(Int, 2^(8sizeof(T)))
+
+    for xi in x
+        @inbounds counts[Int(xi) - typemin(T) + 1] += 1
+    end
+
+    for (i, c) in zip(typemin(T):typemax(T), counts)
+        c == 0 || @inbounds cm[i] = get(cm, i, 0) + c
```

Comment:
I don't think `@inbounds` achieves anything here. FWIW, I think in this loop and loop moving `@inbounds` before the `for` can make the code faster, because (currently) `for xi in x` still includes some bounds checks. Sometimes it prevents the activate of SIMD.

Also, if you really want the maximal performance you could use the same strategy as in `addcounts_dict!` to avoid doing the dict lookup twice. For 16-bit types it could make a small difference.

Reply:
For `Char` I checked on my system `sizeof(Char)` it's actually 4, so 32 bit. So radixsort is the best bet for this.

Reply:
Ah, of course, it's funny how traditional C definitions pollute my mind...